### PR TITLE
Fix duplicate file edge case for deletion and add `remount-hook`

### PR DESCRIPTION
**Pull request by benradf**

### CODE REVIEW

File: src/libstore/local-overlay-store.cc [hunk 1]
```diff
@@ -176,18 +176,42 @@ void LocalOverlayStore::registerValidPaths(const ValidPathInfos & infos)
...
         return;
     }
-    if (pathExists(toUpperPath({path.substr(mergedDir.length())}))) {
-        LocalStore::deleteGCPath(path, bytesFreed);
+
+    StorePath storePath = {path.substr(mergedDir.length())};
+    auto upperPath = toUpperPath(storePath);
+
+    if (pathExists(upperPath)) {
+        std::cerr << "    upper exists" << std::endl;
```
Comment (Ericson2314) [inline src/libstore/local-overlay-store.cc]:
```suggestion
        debug("upper exists: %s", path);
```
or something like that?

File: src/libstore/local-overlay-store.hh [hunk 2]
```diff
@@ -111,11 +118,17 @@ private:
...
+
+    void deleteStorePath(const Path & path, uint64_t & bytesFreed) override;
 
     void optimiseStore() override;
 
     Path toRealPathForHardLink(const StorePath & storePath) override;
+
+    void remountIfNecessary();
+
+    bool _remountRequired = false;
```
Comment (Ericson2314) [inline src/libstore/local-overlay-store.hh]: https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libstore/local-store.hh#L115 Please wrap this with `Sync` like that so we don't risk concurrency issues.

File: src/libstore/local-overlay-store.cc (hunk 1 above)
Comment (benradf) [reply src/libstore/local-overlay-store.cc]: Actually didn't mean to commit this at all! Will fix now.

File: tests/overlay-local-store/delete-inner.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+lowerPath=$(nix-store --store "$storeA" --add delete.sh)
+[[ "$upperPath" = "$lowerPath" ]]
+
+# Check there really are two files with different inodes
+upperInode=$(stat -c %i "$storeBRoot/$upperPath")
+lowerInode=$(stat -c %i "$storeA/$lowerPath")
+[[ "$upperInode" != "$lowerInode" ]]
+
+# Now delete file via the overlay store
+nix-store --store "$storeB&remount-hook=$PWD/remount.sh" --delete "$upperPath"
```
Comment (Ericson2314) [inline tests/overlay-local-store/delete-inner.sh]: https://github.com/NixLayeredStore/nix/actions/runs/5669162468/job/15361297296?pr=11#step:6:3925 huh test failure with this.

File: tests/overlay-local-store/remount.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
```
Comment (Ericson2314) [inline tests/overlay-local-store/remount.sh]:
```suggestion
#!/bin/sh
```
don't have `/usr/bin/env` in namespace. We can get bash by doing `remount.sh.in` (see other such files) but I don't think we need it for this script.

File: src/libstore/local-overlay-store.cc (hunk 1 above)
Comment (benradf) [reply src/libstore/local-overlay-store.cc]: Could be useful for debugging though I suppose :thinking:

**Review by Ericson2314 — APPROVED** [review]

File: tests/overlay-local-store/delete-inner.sh (hunk 3 above)
Comment (benradf) [reply tests/overlay-local-store/delete-inner.sh]: Hmm, because of the `/usr/bin/env` shebang issue you mentioned in the comment below?

**Pull request merged**